casper/src/main/scala/coop/rchain/casper/blocks/merger/ConflictDetectors.scala

```diff
@@ -0,0 +1,308 @@
+                _ <- consumesDestroyedRef
+                      .update(s => s + c.consume)
+              } yield ()
+            }
+
+          })
+      }
+      _ <- fs2.Stream.emits(ps).parJoinUnbounded.compile.toList
+
+      chanToDeployMap <- chanToDeployMapRef.get
+      deployToChanMap <- deployToChanMapRef.get
+
+      producesLinear     <- producesLinearRef.get
+      producesPersistent <- producesPersistentRef.get
+      producesDestroyed  <- producesDestroyedRef.get
+      producesPeeked     <- producesPeekedRef.get
+      consumesLinear     <- consumesLinearRef.get
+      consumesPersistent <- consumesPersistentRef.get
+      consumesDestroyed  <- consumesDestroyedRef.get
+
+    } yield new MergingBranchIndex(
+      vertices = vertices.toSet,
+      chanToDeployMap = chanToDeployMap,
+      producesLinear = producesLinear,
+      producesPersistent = producesPersistent,
+      producesDestroyed = producesDestroyed,
+      producesPeeked = producesPeeked,
+      consumesLinear = consumesLinear,
+      consumesPersistent = consumesPersistent,
+      consumesDestroyed = consumesDestroyed,
+      deploys = deploys,
+      deployToChanMap = deployToChanMap,
+      rejectedDeploys = Set.empty[ProcessedDeploy]
+    )
+
+}
+
+object ConflictDetectors {
+
+  /**
+    * There are 2 cases when the channel is considered conflicting.
+    *
+    *   1. If the same produce or consume is destroyed in COMM in both branches, this might be a race.
+    *
+    *   If event is present in base state, and been destroyed in both branches, its not clear to which branch
+    *   to allocate this event.
+    *   Produce is considered destroyed in COMM if it is not persistent and been consumed without peek.
+    *   Consume is considered destroyed in COMM when it is not persistent.
+    *
+    *   2. Events that are not destroyed in branch's COMMs can lead to potential COMM during merge.
+    *   Concerning events also should be checked for presence in base state, as from event log it cannot be
+    *   deducted if event is created locally of inherited fom base.
+    *
+    * @return set of channels that are conflicting.
+    */
+  def findConflics[F[_]: Concurrent: Log](
+      main: MergingBranchIndex[F],
+      merge: MergingBranchIndex[F],
+      baseDataReader: LazyKeyValueCache[F, Blake2b256Hash, Seq[Datum[ListParWithRandom]]],
+      baseContReader: LazyKeyValueCache[F, Blake2b256Hash, Seq[WaitingContinuation[
+        BindPattern,
+        TaggedContinuation
+      ]]]
+  ): F[Set[Blake2b256Hash]] = {
+
+    // Check #1
+    val biCommedConsume = merge.consumesDestroyed
+      .intersect(main.consumesDestroyed)
+    val biCommedProduce = merge.producesDestroyed
+      .intersect(main.producesDestroyed)
+
+    val competeForBaseEventChecks = biCommedConsume.toList.map(
+      c =>
+        fs2.Stream
+          .eval(for {
+            baseConts <- baseContReader.get(ChannelStoreImpl.continuationKey(c.channelsHashes))
```

Comment:
I just want to say that my old method doesn't have to read from base to know if main and merging competes the same produce(or consume), it can tell from the eventLog. But it can not work with PEAK situation. With our current implementation of peak(consume then produce),  reading base seems unavoidable.